File: Bindings~/python/altunityrunner/alt_unity_port_forwarding.py
import subprocess

from ppadb.client import Client as AdbClient


class AltUnityPortForwarding:
    _client = AdbClient(host="127.0.0.1", port=5037)

    @staticmethod
    def _get_iproxy_path(iproxy_path):
        if not iproxy_path:
            return "iproxy"
        return iproxy_path

    @classmethod
    def _get_device(cls, device_id=""):
        if device_id == "":
            devices = cls._client.devices()
            if len(devices) == 0:
                raise Exception("No device found")
            return devices[0]
        else:
            return cls._client.device(device_id)

    @staticmethod
    def forward_ios(local_port=13000, device_port=13000, device_id="", iproxy_path=""):
        process = None
        iproxy_path = AltUnityPortForwarding._get_iproxy_path(iproxy_path)
        if device_id == "":
            process = subprocess.Popen(
                [iproxy_path, str(local_port), str(device_port)])
        else:
            process = subprocess.Popen(
                [iproxy_path, str(local_port), str(device_port), "-u", device_id])
        return process.pid

    @staticmethod
    def kill_iproxy(pid):
        subprocess.Popen(['kill', str(pid)]).wait()

    @staticmethod
    def kill_all_iproxy_process():
        subprocess.Popen(['killall', 'iproxy']).wait()

    @staticmethod
    def forward_android(local_port=13000, device_port=13000, device_id=""):
        device = AltUnityPortForwarding._get_device(device_id)

        device.forward("tcp:" + str(local_port), "tcp:" + str(device_port))

    @staticmethod
    def remove_forward_android(local_port=13000, device_id=""):
        device = AltUnityPortForwarding._get_device(device_id)
        device.killforward("tcp:" + str(local_port))

    @staticmethod
    def remove_all_forward_android():
        AltUnityPortForwarding._client.killforward_all()

File: Bindings~/java/src/main/java/ro/altom/altunitytester/AltUnityPortForwarding.java
package ro.altom.altunitytester;

import java.nio.file.Paths;

import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import ro.altom.altunitytester.altUnityTesterExceptions.PortForwardingException;

public class AltUnityPortForwarding {
    private static final Logger log = LogManager.getLogger(AltUnityPortForwarding.class);

    public static String getAdbPath(String adbPath) {
        if (adbPath != null && !adbPath.isEmpty())
            return adbPath;

        String androidSdkRoot = System.getenv("ANDROID_SDK_ROOT");

        if (androidSdkRoot != null && !androidSdkRoot.isEmpty()) {
            return Paths.get(androidSdkRoot, "platform-tools", "adb").toString();
        }
        return "adb";
    }

    public static String getIproxyPath(String iproxyPath) {
        if (iproxyPath != null && !iproxyPath.isEmpty())
            return iproxyPath;

        return "iproxy";
    }

    public static void forwardAndroid() {
        forwardAndroid(13000, 13000, "", "");
    }

    public static void forwardAndroid(int localPort) {
        forwardAndroid(localPort, 13000, "", "");
    }

    public static void forwardAndroid(int localPort, int remotePort) {
        forwardAndroid(localPort, remotePort, "", "");
    }

    public static void forwardAndroid(int localPort, int remotePort, String deviceId) {
        forwardAndroid(localPort, remotePort, deviceId, "");
    }

    public static void forwardAndroid(int localPort, int remotePort, String deviceId, String adbPath) {
        adbPath = getAdbPath(adbPath);
        log.debug("Setting up port forward for android local port " + localPort + " remote port: " + remotePort);

        String command;
        if (deviceId.equals(""))
            command = adbPath + " forward tcp:" + localPort + " tcp:" + remotePort;
        else
            command = adbPath + " -s " + deviceId + " forward  tcp:" + localPort + " tcp:" + remotePort;

        try {
            Runtime.getRuntime().exec(command);
            Thread.sleep(1000);
        } catch (Exception ex) {
            throw new PortForwardingException("An exception occured while running command: " + command, ex);
        }

        log.debug("adb forward enabled.");
    }

    public static void removeForwardAndroid() {
        removeForwardAndroid(13000, "", "");
    }

    public static void removeForwardAndroid(int localPort) {
        removeForwardAndroid(localPort, "", "");
    }

    public static void removeForwardAndroid(int localPort, String deviceId) {
        removeForwardAndroid(localPort, deviceId, "");
    }

    public static void removeForwardAndroid(int localPort, String deviceId, String adbPath) {
        adbPath = getAdbPath(adbPath);
        log.debug("Removing android forward localPort: " + localPort + " deviceId: " + deviceId);

        String arguments = "forward --remove tcp:" + localPort;
        if (deviceId != null && !deviceId.isEmpty()) {
            arguments = "-s " + deviceId + " " + arguments;
        }
        String command = adbPath + " " + arguments;

        try {
            Runtime.getRuntime().exec(command);
            Thread.sleep(1000);
        } catch (Exception ex) {
            throw new PortForwardingException("An exception occured while running command: " + command, ex);
        }
        log.debug("Android forward removed...");
    }

    public static void removeAllForwardAndroid() {
        removeAllForwardAndroid("");
    }

    public static void removeAllForwardAndroid(String adbPath) {
        adbPath = getAdbPath(adbPath);

        String command = adbPath + " forward --remove-all";
        try {
            Runtime.getRuntime().exec(command);
            Thread.sleep(1000);
        } catch (Exception ex) {
            throw new PortForwardingException("An exception occured while running command: " + command, ex);
        }
        log.debug("Removed all existing adb forwarding...");
    }

    public static void forwardIos() {
        forwardIos(13000, 13000, "", "");
    }

    public static void forwardIos(int localPort) {
        forwardIos(localPort, 13000, "", "");
    }

    public static void forwardIos(int localPort, int devicePort) {
        forwardIos(localPort, devicePort, "", "");
    }

    public static void forwardIos(int localPort, int devicePort, String deviceId) {
        forwardIos(localPort, devicePort, deviceId, "");
    }

    public static void forwardIos(int localPort, int devicePort, String deviceId, String iproxyPath) {
        iproxyPath = getIproxyPath(iproxyPath);
        String arguments;
        if (deviceId == null || deviceId.isEmpty())
            arguments = localPort + " " + devicePort + "&";
        else
            arguments = localPort + " " + devicePort + " -u " + deviceId + "&";
        String command = iproxyPath + " " + arguments;

        try {
            Runtime.getRuntime().exec(command);
            Thread.sleep(1000);
        } catch (Exception ex) {
            throw new PortForwardingException("An exception occured while running command: " + command, ex);
        }
        log.debug("iproxy forward enabled.");
    }

    public static void killAllIproxyProcess() {
        String command = "killall iproxy";
        try {
            Runtime.getRuntime().exec(command);
            Thread.sleep(1000);
        } catch (Exception ex) {
            throw new PortForwardingException("An exception occured while running command: " + command, ex);
        }
        log.debug("Killed any iproxy process that may have been running...");
    }
}
